### PR TITLE
Reimplement torch::flip based on advanced indexing

### PR DESCRIPTION
**Pull request by andfoy**
## Rationale
This PR improves the performance of `torch::flip` by using `TensorIterator` as the same fashion as using `AdvancedIndexing`. Which means that this implementation is semantically equivalent to indexing a tensor using reverse indices `A[dim0_size - 1:0 ..., dimN_size-1:0, ...]`.

## Benchmark results
The following benchmark compares the runtime of this implementation of `flip` against the current implementation, AdvancedIndexing with reversed indices, as well as OpenCV one. The comparison scenarios consider a 4D tensor `[B, C, H, W]`, where the dimensions flipped correspond to `H` (vertical flip) and `W` (horizontal flip) under float32 and uint8 datatypes. 

The benchmark implementation details can be found in https://github.com/andfoy/flip-benchmarks/blob/main/5_Stable_implementation/benchmarks.py. Additionally, there are correctness tests against the current flip implementation in https://github.com/andfoy/flip-benchmarks/blob/main/5_Stable_implementation/main.cpp, which tests against different layouts, datatypes and contiguous/non-contiguous tensors.
 
The following plots correspond to the means of the runtime of each operator after 100 samples. As it is possible to observe, the latest implementation of flip has a runtime similar to the indexing one. Also, the performance gains are up to 6X under some scenarios.

### Horizontal flip (float)
bokeh_plot

### Horizontal flip (uint8)
bokeh_plot(1)

### Vertical flip (float)
bokeh_plot(2)

### Vertical flip (uint8)
bokeh_plot(3)

cc @fmassa @vfdev-5

### CODE REVIEW

**Comment by wenleix** [comment]
Thanks @andfoy ,  wondering why levering  TensorIterator makes it fast?

File: aten/src/ATen/native/TensorTransformations.cpp [hunk 1]
```diff
@@ -13,80 +13,132 @@ namespace native {
...
+  std::vector<Tensor> indices;
+  int64_t element_size_bytes = input.element_size();
+  for(auto dim: flip_dims) {
+    auto dim_size = input.size(dim);
+    auto index = build_index(input.ndimension(), dim, dim_size);
+    auto stride = input.stride(dim);
+    auto input_index_ptr = index.data_ptr<int64_t>();
+
+    for(int64_t i = 0; i < dim_size; i++) {
+      input_index_ptr[i] = static_cast<int64_t>(dim_size - i - 1) * stride * element_size_bytes;
```
Comment (fmassa) [inline aten/src/ATen/native/TensorTransformations.cpp:60]: I think you could move this logic to `build_index`, as of now `build_index` only returns a non-filled tensor, which is a bit weird.

And the signature of `build_index` could be `Tensor input, int64_t flip_dim`, and you can extract the other information from `input` itself.

File: aten/src/ATen/native/TensorTransformations.cpp [hunk 2]
```diff
@@ -13,80 +13,132 @@ namespace native {
...
+
+  int64_t num_indexers;
+  char** indexers;
+  const int64_t* indexer_strides;
+
+  int64_t get(int64_t idx) {
+    int64_t offset = *(int64_t*)&indexers[0][idx * indexer_strides[0]];
+    for (int j = 1; j < num_indexers; j++) {
+      offset += *(int64_t*)&indexers[j][idx * indexer_strides[j]];
+      // offset += value;
```
Comment (fmassa) [inline aten/src/ATen/native/TensorTransformations.cpp:32]: nit: leftover code

File: aten/src/ATen/native/TensorTransformations.cpp [hunk 3]
```diff
@@ -13,80 +13,132 @@ namespace native {
...
+  return config.build();
+}
+
+template<typename scalar_t>
+void flip_cpu_kernel(TensorIterator& iter) {
+  int ntensor = iter.ntensors();
+  // When launch the index parallel version, set a relative samll grain size less than the INTERNAL::GRAIN_SIZE
+  // to make the whole available thread numbers get more balanced work load and a better cache location.
+  // The grain size here is chosen by the op benchmark to overcome the thread launch overhead
+  const int index_parallel_grain_size = 3000;
```
Comment (fmassa) [inline aten/src/ATen/native/TensorTransformations.cpp:86]: Did we actually benchmark this number? If not, then the comment might be stale. This was taken from the advanced indexing kernel, so it might not apply here. Maybe just point out that we are using the same constant as in the advanced indexing code?

File: aten/src/ATen/native/TensorTransformations.cpp [hunk 4]
```diff
@@ -13,80 +13,132 @@ namespace native {
...
+  }
+
+  int64_t num_indexers;
+  char** indexers;
+  const int64_t* indexer_strides;
+
+  int64_t get(int64_t idx) {
+    int64_t offset = *(int64_t*)&indexers[0][idx * indexer_strides[0]];
+    for (int j = 1; j < num_indexers; j++) {
+      offset += *(int64_t*)&indexers[j][idx * indexer_strides[j]];
```
Comment (fmassa) [inline aten/src/ATen/native/TensorTransformations.cpp:31]: For the future: we could look into specializing this when `num_indexers=1`. It could bring additional performance improvements

**Review by fmassa — COMMENTED** [review]
This generally looks pretty good, thanks!

I've left a few comments.

Also, could you run the performance numbers once more but this time from the PyTorch version that you compiled, just to double-check that under the same compilation flags we get the same speed-up as reported?

Additionally, I it might make sense to see if moving this code to the `native/cpu/` folder would bring speed-ups to the code, as it would be compiled with `-maxv` and `-mavx2` flags, potentially allowing for further compiler optimizations.

File: aten/src/ATen/native/TensorTransformations.cpp [hunk 5]
```diff
@@ -13,80 +13,132 @@ namespace native {
...
+  };
+
+  iter.for_each(loop, index_parallel_grain_size);
+}
+
+Tensor flip_cpu(const Tensor& self, IntArrayRef dims) {
+  auto input = self;
+
+  std::vector<int64_t> flip_dims;
+  for(int64_t dim: dims) {
```
Comment (ngimel) [inline aten/src/ATen/native/TensorTransformations.cpp:105]: you don't check here that `dim` is in range, and you don't check that there are no repeating dims (previously it was done by `dim_list_to_bitset`).

Comment (fmassa) [reply aten/src/ATen/native/TensorTransformations.cpp:105]: Yes, good catch! I think we should just use the `dim_list_to_bitset` as before but to build the indices from there

File: aten/src/ATen/native/TensorTransformations.cpp (hunk 3 above)
Comment (andfoy) [reply aten/src/ATen/native/TensorTransformations.cpp:86]: No, we didn't benchmark this number! I'll put the corresponding comment there

**Comment by andfoy** [comment]
These are the benchmark results for commit 0b63646 against the current `torch.flip` implementation. The comparison was done by declaring both implementations on the same PyTorch compilation on https://github.com/pytorch/pytorch/compare/master...andfoy:benchmark_flip

### Horizontal flip (float)
bokeh_plot(4)


### Horizontal flip (uint8)
bokeh_plot(5)


### Vertical flip (float)
bokeh_plot(6)


### Vertical flip (uint8)
bokeh_plot(7)

**Comment by andfoy** [comment]
> Wondering why levering TensorIterator makes it fast?
 
@wenleix  My guess here would be that by precomputing the indices to flip, the runtime due to this loop is removed:

```c++
for (int64_t d = 0; d < total_dims; d++) {
      int64_t temp = cur_indices;
      cur_indices = cur_indices / stride_contiguous_v[d];
      rem = temp - cur_indices * stride_contiguous_v[d];
      dst_offset += flip_dims_b[d] ? (sizes_v[d] - 1 - cur_indices) * strides_v[d] : cur_indices * strides_v[d];
      cur_indices = rem;
}
```

However, I could be wrong here

**Comment by ngimel** [comment]
@andfoy what do you think of this approach https://github.com/pytorch/pytorch/compare/master...ngimel:flip?expand=1 where TI is used directly, without indexing tensors?
The advantage is that it can be very easily extended to cuda too. Perf benchmarks comparing to existing flip:
Before:
```
[-------------------- flip -------------------]
                       |   dim=1    |   dim=2  
1 threads: ------------------------------------
      (7, 112, 3)      |      80.4  |      73.2
      (28, 28, 3)      |      73.4  |      74.0
      (112, 7, 3)      |      71.6  |      74.5
      (8, 2048, 3)     |    1559.4  |    1501.0
      (128, 128, 3)    |    1487.5  |    1488.6
      (2048, 8, 3)     |    1489.0  |    1480.9
      (5, 102400, 3)   |   46704.5  |   46615.1
      (800, 640, 3)    |   46453.5  |   46702.9
      (128000, 4, 3)   |   46944.7  |   47343.2
      (4, 196608, 3)   |   72009.5  |   73965.6
      (1024, 768, 3)   |   72100.0  |   73281.6
      (262144, 3, 3)   |   70512.6  |   72408.6
      (16, 129600, 3)  |  189054.7  |  201812.7
      (1920, 1080, 3)  |  184979.7  |  224558.0
      (230400, 9, 3)   |  197751.2  |  195470.6
```
After
```
[------------------ flip -----------------]
                       |  dim=1   |  dim=2 
1 threads: --------------------------------
      (7, 112, 3)      |     3.7  |     4.0
      (28, 28, 3)      |     4.1  |     4.3
      (112, 7, 3)      |     4.8  |     5.0
      (8, 2048, 3)     |    20.1  |    27.6
      (128, 128, 3)    |    22.4  |    35.2
      (2048, 8, 3)     |    41.4  |    50.4
      (5, 102400, 3)   |   628.8  |   731.0
      (800, 640, 3)    |   651.5  |   939.0
      (128000, 4, 3)   |  1698.2  |  2766.2
      (4, 196608, 3)   |  1130.8  |  1145.4
      (1024, 768, 3)   |  1030.7  |  1643.5
      (262144, 3, 3)   |  3464.7  |  5177.9
      (16, 129600, 3)  |  3583.6  |  4012.7
      (1920, 1080, 3)  |  4049.7  |  4525.1
      (230400, 9, 3)   |  5824.7  |  6371.6
```
Benchmarking script:
<details><summary>CLICK ME</summary>
<p>

```
import torch
from torch.utils.benchmark import Timer
from torch.utils.benchmark import Compare
sizes = [
    (7, 112, 3),
    (28, 28, 3),
    (112, 7, 3),

    (8, 2048, 3),
    (128, 128, 3),
    (2048, 8, 3),

    (5, 102400, 3),
    (800, 640, 3),
    (128000, 4, 3),

    (4, 196608, 3),
    (1024, 768, 3),
    (262144, 3, 3),

    (16, 129600, 3),
    (1920, 1080, 3),
    (230400, 9, 3),

    # (16, 518400, 3),
    # (3840, 2160, 3),
    # (921600, 9, 3),
]
results = []

for size in sizes:

    H, W, C = size


    inp = torch.rand(C, H, W)
    #print(inp.size())
    t1 = Timer(stmt = "torch.flip(inp, [1])", sub_label=f"{size}", description="dim=1", label="flip", globals=globals())
    t2 = Timer(stmt = "torch.flip(inp, [2])", sub_label=f"{size}", description="dim=2", label="flip", globals=globals())
    timers = [t1,t2]
    for t in timers:
        results.append(
            t.blocked_autorange()
        )

comparison=Compare(results)
comparison.print()


```
</p>
</details>

**Comment by andfoy** [comment]
Since it can scale to CUDA easily, and the changes are way more simpler than the ones proposed, I think it is a good option. So in this order of ideas, basically a loop over the dimensions to flip should be called before calling the actual kernel?

```c++
for(int64_t i = 0; i < total_dims; i++) {
   if(flip_dims_b[i]) {
      iter.flip_strides(0, i);
   }
}
```

**Comment by ngimel** [comment]
No, `flip_strides` has to be called only once, and it's implementation in TensorIterator flips all the necessary dimensions. I agree that if we could do it in a loop like you propose, it's conceptually cleaner, but the reason it has to be done all at once is after TI is built, the dims that are being flipped are no longer the dims that were originally specified, because TensorIterator coalesces dimensions that it can view as one larger dim. 
Imagine there's a 3d tensor where you want to flip the last dim. If input is contiguous, TensorIterator will know that it can;t collapse the last dimension because it will be flipped, but it will collapse first 2 dimensions, and will view the tensor as a 2d tensor (size0*size1, size2). So, when flipping strides, you should no longer flip the 2nd (0-based) stride, you need to flip the 1st! Luckily, we are sending a dummy tensor to TensorIterator that tracks which dimensions actually have to be flipped even after coalescing. 
The code I'm proposing is very sparsely tested, so I won't be surprised if there are bugs, don't hold it against me :-)

**Comment by andfoy** [comment]
Thanks for the clarification @ngimel! I'll do a run of your changes against the tests that I have on the other repo. But before continuing with this process, are you okay with adding additional API changes into TensorIterator?

**Comment by fmassa** [comment]
@ngimel I like your approach with changing TensorIterator directly, but I wonder given how widely used it is if it would be ok to extend it's API for a single function to use it?

**Comment by ngimel** [comment]
It's not ideal, but TI already has "unsafe_replace_operand" that has pretty niche use. Alternatively, TI can expose setter APIs for strides (with the caveat that you should know what you are doing), and then this particular stride manipulation can be done in flip implementation.

**Comment by fmassa** [comment]
@ngimel I think that while modifying TI directly can speed things up even further and have a nice path for optimizing the CUDA counter-part as well, I think that making modifications directly into TensorIterator might be something we should consider more carefully given the number of customers it has.

That being said, I would propose we move forward with this PR in its current state, so that we can at least get some nice improvements to the CPU implementation of flip without intrusive changes to TI.

We can always revisit it back in the future to extend TensorIterator with other functionality.

Thoughts?

**Comment by ngimel** [comment]
Conceptually, I don't think inverting strides is any more invasive than restriding we are currently doing for indexing operation. The reason we can't just do restriding here is that negative strides are not supported for tensors. 
But I agree, we can get this PR in in its current form, and optimize later.

**Comment by andfoy** [comment]
I have a question regarding the quantized call for the new flip kernel, should I duplicate this code under `quantized/cpu`, or should I copy it back to `TensorTransformations.cpp`, while we also remove `FlipKernel.cpp`?

**Comment by ngimel** [comment]
Did you verify that duplicating code under native/cpu actually improves performance compared to keeping in in just native?

**Comment by andfoy** [comment]
@ngimel, let me check the performance comparison; if the performance is at par or if the gains are marginal, then I'll keep the kernel under `TensorTransforms`

**Comment by andfoy** [comment]
I checked the benchmark results, and the differences are not significant, which means that we can leave the kernel in `TensorTransformations`

**Comment by andfoy** [comment]
The error in ROCm seems to be unrelated to this PR

**Comment by andfoy** [comment]
@ngimel @fmassa @wenleix  This one is ready for a final review

File: aten/src/ATen/native/TensorTransformations.cpp [hunk 6]
```diff
@@ -13,82 +13,146 @@ namespace native {
...
-        int64_t temp = cur_indices;
-        cur_indices = cur_indices / stride_contiguous_v[d];
-        rem = temp - cur_indices * stride_contiguous_v[d];
-        dst_offset += flip_dims_b[d] ? (sizes_v[d] - 1 - cur_indices) * strides_v[d] : cur_indices * strides_v[d];
-        cur_indices = rem;
-      }
-      out_tensor_d[i] = in_tensor_d[dst_offset];
+void flip_cpu_kernel(TensorIterator& iter) {
+  int ntensor = iter.ntensors();
+  // When launch the index parallel version, set a relative samll grain size
```
Comment (vfdev-5) [inline aten/src/ATen/native/TensorTransformations.cpp:92]:
```suggestion
  // When launch the index parallel version, set a relative small grain size
```

File: aten/src/ATen/native/TensorTransformations.cpp [hunk 7]
```diff
@@ -13,82 +13,146 @@ namespace native {
...
+  // less than the INTERNAL::GRAIN_SIZE to make the whole available thread
+  // numbers get more balanced work load and a better cache location. The grain
+  // size here is chosen by the op benchmark to overcome the thread launch
+  // overhead. This value was taken from the AdvancedIndexing kernel.
+  const int index_parallel_grain_size = 3000;
```
Comment (vfdev-5) [inline aten/src/ATen/native/TensorTransformations.cpp:97]: @andfoy any gains using this value vs default one ?

Comment (andfoy) [reply aten/src/ATen/native/TensorTransformations.cpp:97]: Let me check!

Comment (andfoy) [reply aten/src/ATen/native/TensorTransformations.cpp:97]: See https://github.com/pytorch/pytorch/pull/56713#issuecomment-837354845

File: aten/src/ATen/native/TensorTransformations.cpp [hunk 8]
```diff
@@ -13,81 +13,145 @@ namespace native {
 
 constexpr size_t dim_bitset_size = 64;
 
+Tensor build_index(Tensor input, int64_t flip_dim) {
```
Comment (fmassa) [inline aten/src/ATen/native/TensorTransformations.cpp:16]: Can we add all these internal functions inside an anonymous namespace? Given that those names are very generic there could potentially be conflicts with other files.

See for example how it's done in https://github.com/pytorch/pytorch/blob/18edb77a28406e97d6b571f7761a50923293bf64/aten/src/ATen/native/cpu/IndexKernel.cpp#L12

So that all the `build_index`, `build_indices_loop` , `make_index_iterator` and `Indexer` are in the private namespace.

Thoughts?

**Review by fmassa — COMMENTED** [review]
This looks good to me, thanks!

I only have one comment which I think could be good to have, otherwise I think this is good for merge.

Let me know what you think

**Review by fmassa — APPROVED** [review]
Thanks!
